csr/csr.go

```diff
@@ -92,22 +96,34 @@ func VerifyCSR(ctx context.Context, csr *x509.CertificateRequest, maxNames int,
 	return nil
 }
 
-// normalizeCSR deduplicates and lowers the case of dNSNames and the subject CN.
-// It will also hoist a dNSName into the CN if it is empty.
-func normalizeCSR(csr *x509.CertificateRequest) {
-	if csr.Subject.CommonName == "" {
-		var forcedCN string
-		// Promote the first SAN that is less than maxCNLength (if any)
-		for _, name := range csr.DNSNames {
-			if len(name) <= maxCNLength {
-				forcedCN = name
-				break
-			}
+type names struct {
+	SANs []string
+	CN   string
+}
+
+// NamesFromCSR deduplicates and lower-cases the Subject Common Name and Subject
+// Alternative Names from the CSR. It guarantees that the SANs include the CN.
+// It also enforces various conditions on the CN, based on feature flags.
+func NamesFromCSR(csr *x509.CertificateRequest) names {
+	sans := csr.DNSNames
+	if csr.Subject.CommonName != "" {
+		sans = append(sans, csr.Subject.CommonName)
+	}
+	sans = core.UniqueLowerNames(sans)
+
+	if !features.Enabled(features.SetCommonName) {
+		return names{SANs: sans}
+	}
+
+	if csr.Subject.CommonName != "" {
+		return names{SANs: sans, CN: strings.ToLower(csr.Subject.CommonName)}
+	}
+
+	for _, name := range sans {
```

Comment:
This old (removed) comment helped me figure out what this for loop was doing. I think it should come back.

```suggestion
	// Promote the first SAN that is less than maxCNLength (if any)
	for _, name := range sans {
```

Reply:
Done!